### PR TITLE
[Frontend] Tool calling parser for Granite 3.0 models

### PR DESCRIPTION
**Pull request by maxdebayser**
This PR adds a tool calling parser for `ibm-granite/granite-3.0-8b-instruct`. The smaller models in 
the [Granite 3.0 Language Models](https://huggingface.co/collections/ibm-granite/granite-30-models-66fdb59bbb54785c3512114f) models collection pass some of the tests..

cc: @njhill

### CODE REVIEW

File: docs/source/serving/openai_compatible_server.md [hunk 1]
```diff
@@ -157,7 +157,7 @@ vLLM will use guided decoding to ensure the response matches the tool parameter
 To enable this feature, you should set the following flags:
 * `--enable-auto-tool-choice` -- **mandatory** Auto tool choice. tells vLLM that you want to enable the model to generate its own tool calls when it 
 deems appropriate.
-* `--tool-call-parser` -- select the tool parser to use - currently either `hermes` or `mistral` or `llama3_json` or `internlm`. Additional tool parsers 
+* `--tool-call-parser` -- select the tool parser to use - currently either `hermes`, `mistral`, `llama3_json`, `internlm` or `granite-20b-fc`. Additional tool parsers
```
Comment (njhill) [inline docs/source/serving/openai_compatible_server.md:160]: @maxdebayser should `granite` also be added here?

File: docs/source/serving/openai_compatible_server.md [hunk 2]
```diff
@@ -219,6 +222,20 @@ it works better with vLLM.
...
 
+#### IBM Granite
+
+Supported models:
+* `ibm-granite/granite-20b-functioncalling`
+
+Flags: `--tool-call-parser granite-20b-fc`
+`examples/tool_chat_template_granite_20b_fc.jinja`: this is a modified chat template from the original on Huggingface, which is not vLLM compatible. It blends function description elements from the Hermes template and follows the same system prompt as "Response Generation" mode from [the paper](https://arxiv.org/abs/2407.00121). Parallel function calls are supported.
+
+* `ibm-granite/granite-8b-instruct`
```
Comment (njhill) [inline docs/source/serving/openai_compatible_server.md:233]:
```suggestion
* `ibm-granite/granite-3.0-8b-instruct`
```

File: docs/source/serving/openai_compatible_server.md [hunk 3]
```diff
@@ -219,6 +222,20 @@ it works better with vLLM.
 
 Recommended flags: `--tool-call-parser llama3_json --chat-template examples/tool_chat_template_llama3_json.jinja`
 
+#### IBM Granite
+
+Supported models:
+* `ibm-granite/granite-20b-functioncalling`
+
+Flags: `--tool-call-parser granite-20b-fc`
+`examples/tool_chat_template_granite_20b_fc.jinja`: this is a modified chat template from the original on Huggingface, which is not vLLM compatible. It blends function description elements from the Hermes template and follows the same system prompt as "Response Generation" mode from [the paper](https://arxiv.org/abs/2407.00121). Parallel function calls are supported.
```
Comment (njhill) [inline docs/source/serving/openai_compatible_server.md:231]: Is `--chat-template` missing here? (see llama above for example)

File: docs/source/serving/openai_compatible_server.md [hunk 4]
```diff
@@ -219,6 +222,20 @@ it works better with vLLM.
...
+
+* `ibm-granite/granite-8b-instruct`
+
+Flags: `--tool-call-parser granite`
+`examples/tool_chat_template_granite.jinja`: this is a modified chat template from the original on Huggingface. Parallel function calls are supported.
```
Comment (njhill) [inline docs/source/serving/openai_compatible_server.md:236]: Is `--chat-template` missing here?

File: docs/source/serving/openai_compatible_server.md [hunk 5]
```diff
@@ -230,6 +247,15 @@ Known issues:
...
 
+#### IBM Granite
+
+Supported models:
+* `ibm-granite/granite-20b-functioncalling`
+
+Flags: `--tool-call-parser granite-20b-fc`
+`examples/tool_chat_template_granite_20b_fc.jinja`: this is a modified chat template from the original on Huggingface, which is not vLLM compatible. It blends function description elements from the Hermes template and follows the same system prompt as "Response Generation" mode from [the paper](https://arxiv.org/abs/2407.00121). Parallel function calls are supported.
+
+
```
Comment (njhill) [inline docs/source/serving/openai_compatible_server.md:258]: Duplicate, should be removed?

File: docs/source/serving/openai_compatible_server.md [hunk 6]
```diff
@@ -219,6 +222,20 @@ it works better with vLLM.
 
 Recommended flags: `--tool-call-parser llama3_json --chat-template examples/tool_chat_template_llama3_json.jinja`
 
+#### IBM Granite
+
+Supported models:
+* `ibm-granite/granite-20b-functioncalling`
```
Comment (njhill) [inline docs/source/serving/openai_compatible_server.md:228]: Probably better to swap the order of these and move the granite 3.0 one to the top

**Review by njhill — COMMENTED** [review]
Thanks @maxdebayser!

**Comment by maxdebayser** [comment]
This is ready for review.

cc: @njhill @wseaton

File: vllm/entrypoints/openai/tool_parsers/granite_tool_parser.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+                    prev_arguments = self.prev_tool_call_arr[
+                        self.current_tool_id].get("arguments")
+
+                    argument_diff = None
+                    if is_complete[self.current_tool_id]:
+                        argument_diff = cur_args_json[sent:]
+                    elif prev_arguments:
+                        prev_args_json = json.dumps(prev_arguments)
+                        if cur_args_json != prev_args_json:
+
```
Comment (njhill) [inline vllm/entrypoints/openai/tool_parsers/granite_tool_parser.py:204]: nit: remove this empty line?

**Review by njhill — COMMENTED** [review]
Thanks @maxdebayser! I did a first pass and left some comments

**Comment by njhill** [comment]
cc @K-Mistele in case you'd like to take a look too

**Comment by K-Mistele** [comment]
will take another look!

**Review by njhill — APPROVED** [review]
Thanks again @maxdebayser. I'll hold off for any comments from @K-Mistele before merging.

**Comment by K-Mistele** [comment]
Giving it a final once-over :) 

Out of curiosity it looks like the model's default context is 4096; is there any way to scale this if I want to test longer-context tool calls? no worries if there's not a good way to do with with vLLM

**Comment by K-Mistele** [comment]
Looks great to me! Seems pretty robust, and passing tests on my machine :)

**Comment by K-Mistele** [comment]
Looks like the only failing tests were AMD-related which was expected last I checked. cc @mgoin @DarkLight1337

**Comment by DarkLight1337** [comment]
> Looks like the only failing tests were AMD-related which was expected last I checked. cc @mgoin @DarkLight1337

The test is not failing on main branch, so I think it's introduced by this PR. Do tell me if I'm mistaken though.

**Comment by K-Mistele** [comment]
> > Looks like the only failing tests were AMD-related which was expected last I checked. cc @mgoin @DarkLight1337
> 
> The test is not failing on main branch, so I think it's introduced by this PR. Do tell me if I'm mistaken though.

Oops, missed that. I just remembered the last time I did one there were some expected to fail, but I'm sure you're right :) 

Seems related to the FP8 granite 20B quantization (`mbayser/granite-20b-functioncalling-FP8-KV`) that's being used:

https://buildkite.com/vllm/ci-aws/builds/10840#01930511-9576-45cb-aebb-e237d5f07c9b/974-5501

AMD supports FP8 in ROCM > 6.2. Could also be an OOM issue since this is a 20B that we had issues fitting into other CI GPUs; or could be related to CPU offloading. 

Whatever the case, seems more like a model+configuration+hardware compatibility issue rather than a code issue. Is there any way to disable this particular model for AMD tests, and would that be an acceptable solution?

**Comment by DarkLight1337** [comment]
I think this is alright. @njhill any thoughts?

**Comment by maxdebayser** [comment]
I've pushed a change to disable skip the `granite-20b-functioncalling` test on AMD. Let's see if it passes now :crossed_fingers:

**Comment by maxdebayser** [comment]
> Out of curiosity it looks like the model's default context is 4096; is there any way to scale this if I want to test longer-context tool calls? no worries if there's not a good way to do with with vLLM

I'm not sure, but as we add more models and tests, we might have to get the max sequence length for each model and skip the tests that require longer ones.

**Comment by njhill** [comment]
Thanks again @maxdebayser @K-Mistele @DarkLight1337!

**Pull request merged**